unliftio/src/UnliftIO/Internals/Async.hs

```diff
@@ -0,0 +1,834 @@
+              -- synchronous exceptions, so, any exception that the try above would catch
+              -- must be an async exception.
+              -- We were killed by an async exception, do nothing.
+              Left (_ :: E.SomeException) -> pure ()
+              -- Child thread died, propagate it
+              Right (Left e)              -> void $ tryPutTMVar excVar0 e
+              -- Successful result from one of the children
+              Right (Right res)           -> putTMVar resVar res
+
+          -- And kill all of the threads
+          for_ workerTids $ \tids' ->
+            -- NOTE: Replacing A.AsyncCancelled with KillThread as the
+            -- 'A.AsyncCancelled' constructor is not exported in older versions
+            -- of the async package
+            -- for_ (tids' []) $ \workerTid -> E.throwTo workerTid A.AsyncCancelled
+            for_ (dlistToList tids') $ \workerTid -> C.killThread workerTid
+
+        pure ( readTMVar resVar
+             , helperTid `dlistCons` dlistConcatAll workerTids
+             )
+
+  excVar <- newEmptyTMVarIO
+  (getRes, tids0) <- go excVar f0
+  let tids = dlistToList tids0
+      tidCount = length tids
+      allDone count =
+        if count > tidCount
+          then error ("allDone: count ("
+                      <> show count
+                      <> ") should never be greater than tidCount ("
+                      <> show tidCount
+                      <> ")")
+          else count == tidCount
+
+  -- Automatically retry if we get killed by a
+  -- BlockedIndefinitelyOnSTM. For more information, see:
+  --
+  -- + https:\/\/github.com\/simonmar\/async\/issues\/14
+  -- + https:\/\/github.com\/simonmar\/async\/pull\/15
+  --
+  let autoRetry action =
+        action `E.catch`
+        -- NOTE: This @case@ is important as it will display new warnings in the
+        -- scenario where the `BlockedIndefinitelyOnSTM` type has new
+        -- constructor values
+        \case E.BlockedIndefinitelyOnSTM -> autoRetry action
+
+  -- Restore the original masking state while blocking and catch
+  -- exceptions to allow the parent thread to be killed early.
+  res <- E.try $ restore $ autoRetry $ atomically $
+         (Left <$> readTMVar excVar) <|>
+         (Right <$> getRes)
+
+  count0 <- atomically $ readTVar resultCountVar
+  unless (allDone count0) $ do
+    -- Kill all of the threads
+    -- NOTE: Replacing A.AsyncCancelled with KillThread as the
+    -- 'A.AsyncCancelled' constructor is not exported in older versions
+    -- of the async package
+    -- for_ tids $ \tid -> E.throwTo tid A.AsyncCancelled
+    for_ tids $ \tid -> C.killThread tid
+
+    -- Wait for all of the threads to die. We're going to restore the
+    -- original masking state here, just in case there's a bug in the
+    -- cleanup code of a child thread, so that we can be killed by an
+    -- async exception.
+    -- TODO: Why do we want to allow to be killed, why is not ok to hang in this situation?
```

Comment:
@snoyberg can you give me a bit of context around why we picked this solution above all others?

Reply:
There's not much more to say than what I did already. It's a trade-off: do we want to hang indefinite in the case of buggy cleanup code, or allow a second async exception to take the whole thing down? It somewhat reminds me of the PID1 situation where we'll send a SIGTERM and then, a few seconds later, a SIGKILL, with the theory that we don't want to allow the system to indefinitely hang during shutdown because of one bad process.